### ISSUE TITLE
FR: add support for other read-it-later services (omnivore,shiori)

### ISSUE THREAD

**Issue by Amolith** [koreader/koreader#10406]
[Omnivore](https://omnivore.app) ([GitHub](https://github.com/omnivore-app/omnivore)) is an open source read-it-later service similar to wallabag, which KOReader already has support for. I'd like to see support added for Omnivore as well; its UI is much more pleasant than wallabag's and they have a promising roadmap.

If this would be welcome as a PR, I'll take a look when I have time over the summer.

- [Vague API docs on their website](https://docs.omnivore.app/integrations/api.html)
- [GraphQL schema on GitHub](https://github.com/omnivore-app/omnivore/blob/main/packages/api/src/schema.ts)

**Comment by Frenzie** [comment]
If it works sufficiently similar maybe the Wallabag plugin can be generalized. I'd be less enthused about having another plugin that's extremely similar but different, anyway. ;-)

**Comment by pazos** [comment]
A bit dissapointing https://github.com/omnivore-app/omnivore#how-to-deploy-to-your-own-server

**Comment by aqxa1** [comment]
There's also [shiori](https://github.com/go-shiori/shiori) which is much faster than wallabag on my system (at least with the official docker container) and deployment is simple.

**Comment by Amolith** [comment]
Looking around some more, neither Omnivore nor Shiori export EPUB files. Wallabag definitely does and I'm guessing that the current KOReader plugin just lets wallabag do the hard conversion work and it downloads the result.

Adding Omnivore/Shiori support would make the plugin much more complex because it would have to do all that conversion itself. I don't have the time or motivation to implement this, so I'll just stick with wallabag for now ^^'

**Comment by aqxa1** [comment]
A PR for [epub generation](https://github.com/go-shiori/shiori/pull/623) is close to being merged to shiori. So hopefully it becomes realistic once that happens.

**Comment by Frenzie** [comment]
Creating an EPUB in KOReader isn't much work — or rather, the work's already been done before.

That being said, in those linked Shiori API docs I don't even see how to get the content, just the URL? Now that could complicate matters significantly, given how authentication can also be an important part of these systems, cf. `https://github.com/wallabag/wallabag/pull/2317`.

**Comment by Frenzie** [comment]
Other than actually getting content in any form at all I don't see anything obviously missing in the API docs.

For PDF you'd generally want to download the actual PDF instead of some text extracted EPUB.

**Comment by Frenzie** [comment]
Edit would ideally also include read status though; I don't see that listed.

**Comment by pazos** [comment]
@Monirzadeh: you'll need to populate `sorting_hint` with the string that defines a submenu. You can get all the strings in https://github.com/koreader/koreader/blob/master/frontend/ui/elements/reader_menu_order.lua and related files.

**Comment by pazos** [comment]
> @pazos i add `sorting_hint = "tools",` and it move to the tools menu but after `More tools` in second page. can i move that to the first page? for example after wallabag? ![3](https://private-user-images.githubusercontent.com/25131576/298901352-44aa0fbf-a868-445f-b4ad-adb52e96f28e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYwMTYyMzIsIm5iZiI6MTcwNjAxNTkzMiwicGF0aCI6Ii8yNTEzMTU3Ni8yOTg5MDEzNTItNDRhYTBmYmYtYTg2OC00NDVmLWI0YWQtYWRiNTJlOTZmMjhlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIzVDEzMTg1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyNDgzM2U2N2I0MWIwMTM4NDhhMDFjYjc1Y2MzOWI1ZDEzZjE5MTM2Yzg1YzcxMWY0NmE0ODA0NDJjNzA5MGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4l0_pQBM8wF0o7gEbmywrDVfjmWFzCW1SfrN2xgDhq4)
> 
> note: i add shiori in https://github.com/koreader/koreader/blob/master/frontend/ui/elements/reader_menu_order.lua after wallabag too, but it is not work

Your guess is good. If you want to place the menu entry at a fixed position you need to edit the menu file and ignore sorting hint.

Editing reader menu order will affect the menu of the reader. You need to do ir also on the file manager menu order.

**Comment by pazos** [comment]
@Monirzadeh: could you explain the features of the plugin?

Also, key features of shiaori. Is a drop-in replacement for wallabag? advantages, features...

> what is the workflow for adding shiori plugin to koreader?

Releasing on your own is quick and easy. Submitting a PR here requires a long review. I think it worths our time, in this case :)

**Comment by pazos** [comment]
That looks awesome!

Is it self-contained on `shiaori.koplugin` or requires other changes too?

Feel free to open a wip PR, maybe with a todo list for the things that you want to be done.

**Comment by Frenzie** [comment]
That's debatable. The plugin presumably needs some settings regardless (e.g., a limit for article age or number, whether you want read = delete, etc.), so it's not a stretch to put the authorization in there too, quite the opposite in our opinion.

But we're probably speaking in the abstract a bit too much atm.

**Comment by pazos** [comment]
@Monirzadeh:

If you want to add full support for shiori you'll need to make different PRs.

First focus on the plugin.
Then, you can add shiori as an export target.

The cloud-storage thing is a bit nasty to merge stuff in. Hopefully we should be able to move stuff out. We don't have an account manager yet, but if you want to commit one make sure it is separated from cloudstorage

If you need a cloudstorage view for shiori then feel free to submit those changes in yet another PR. At some point we should refactor `frontend/apps/cloudstorage` as a filemanager module that displays virtual files.

**Comment by tomtom800** [comment]
There is now an [Omnivore plugin](https://github.com/fairlygood/omnivore.koplugin) which works well in my brief testing. It would good if this led to official support for Omnivore. It’s a good read later service under active development.

**Comment by pazos** [comment]
> There is now an [Omnivore plugin](https://github.com/fairlygood/omnivore.koplugin) which works well in my brief testing.

That's great! Thanks for the heads up!

> It would good if this led to official support for Omnivore.

It would not :). We're very proud of third party development. Notable examples are https://github.com/stelzch/zotero.koplugin, https://github.com/Ajatt-Tools/anki.koplugin or https://github.com/drewbaumann/AskGPT.

We're also very happy of development and support happening elsewhere :)

> It’s a good read later service under active development and much more user friendly than Wallabag.

I disagree :). It isn't still a free service. You cannot self host yet (as stated in https://github.com/koreader/koreader/issues/10406#issuecomment-1540091406) so, while it may be free software you're relying on somebody's else computer.

Even worse than that!. It seems to output documents as html. Wallabag sends epub. That's a matter of preference, I guess. But epub capabilities on KOReader are overwhelming compared to html support.

**Comment by tomtom800** [comment]
I understand the advantages of Wallabag, but I don't think I'm that unusual in being a Koreader user who doesn't want (or have the skills) to self-host a read later service - so for that sort of service I'm always going to be using "someone else's computer". I paid for the hosted Wallabag service so that I could use it with Koreader, but I didn't have a great experience - quite of lot articles didn't save or format correctly, and my emails to their support were not answered. I tried to troubleshoot it myself but I really struggled with the complexity of that process and eventually gave up. 

Omnivore, by contrast, has been easy to use throughout and I've had a near 100% success rate in saving articles.  I find the html formatting via Omnivore to be completely acceptable for my use case (which I acknowledge is mainly short articles). 

My point is that I think there are advantages in building in a second service that gives new users a less complex experience.

**Comment by pazos** [comment]
Thanks for the feedback, @tomtom800.

> My point is that I think there are advantages in building in a second service that gives new users a less complex experience.

I 100% agree. But that doesn't need to happen here.
It should happen where the author actually use the service and related plugin.

Upstreaming 3rd party plugins won't happen unless the original author agrees and wants to commit effort porting and supporting the plugin.

In that regard, plugins that come from contributors, like @Monirzadeh, are safer to upstream, as the contribution history is there as a seal of trust :)

But in the case of **Read it Later** kind of services and pervious comments: IMHO shiori looks much better. So hopping for that one upstream. And omnivore as 3rd party :)

We actually provide a very good service for 3rd party plugins, not by design but by defect :). Plugins (and runtime patches) allow to do exactly the same as core. The only thing missing in 3rd party plugins is the translation framework.

**tl;dr:** the only thing the maintainer gains upstreaming is the translation framework. The maintainer loses peace of mind and control over the code/repo :/. Upstream gains FUD over future support :d

**Comment by tomtom800** [comment]
Shiori looks great, but it's another self-hosted solution. Instructions that begin "first install Docker" will lose basically every novice user.

Anyway, thanks for taking the time to give such detailed replies - I appreciate it.

**Comment by hius07** [comment]
> if it's possible to highlight with this new plugin

No, it just saves a web page from your Omnivore library to the device as html file.

@tomtom800 are you in contact with the author? To bring up some issues.

**Comment by tomtom800** [comment]
> @tomtom800 are you in contact with the author? To bring up some issues.

No, but I linked to the GitHub page further up the thread. 

I’ve just had the email that @AaronKnowls references. Sadly it does sound as though Omnivore will close in a couple of weeks so I don’t think there’s any point in developing the plugin any further.

**Comment by hius07** [comment]
> GitHub page

I saw, the author turned issues off. Thanks anyway.

**Comment by tomtom800** [comment]
There is also [a reddit thread.](https://www.reddit.com/r/koreader/comments/1g3j29z/omnivore_plugin/)

**Comment by pazos** [comment]
> would anyone be open to support an integration for Readeck or Linkding as well?

All of these read-it-later services can be implemented as standalone plugins and supported elsewhere. That's the preferred way :)

If/when one of those plugins is available feel free to let us know here :)
